compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2213,16 +2213,38 @@ class Typer extends Namer
             traverse(xtree :: rest, facts)
           case none =>
             import untpd.modsDeco
-            val ctx1 = if (ctx.explicitNulls) {
-              mdef match {
-                case mdef: untpd.ValDef if ctx.owner.is(Method) && !mdef.mods.isOneOf(Lazy | Implicit) =>
-                  ctx.fresh.addFlowFacts(facts)
-                case _ => ctx
+            // Force completer to use calling context (as opposed to the creation context)
+            // to complete itself. This approach is used so that flow typing can handle `ValDef`s
+            // that appear within a block.
+            //
+            // Example:
+            // Suppose we have a block containing
+            // 1. val x: String|Null = ???
+            // 2. if (x == null) throw NPE
+            // 3. val y = x
+            //
+            // We want to infer y: String on line 3, but if the completer for `y` uses its creation context,
+            // then we won't have the additional flow facts that say that `y` is not null.
+            //
+            // The solution is to use the context containing more information about the statements above.
+            val ctx2 = if (ctx.explicitNulls && ctx.owner.is(Method)) {
+              val ctx1 = ctx.fresh.addFlowFacts(facts)
+              // we cannot use mdef.symbol to get the symbol of the tree here
+              // since the tree has not been completed and doesn't have a denotation
+              mdef.getAttachment(SymOfTree) match {
+                case Some(sym) =>
+                  sym.infoOrCompleter match {
+                    case completer: Namer#Completer =>
+                      completer.completeInContext(sym, ctx1)
+                    case _ =>
```

Comment:
Please add a comment explaining that if the completer has already been forced, then it must be because there is a forward-reference to the definition in the program. In that case, we won't use the current context and flow-typing will not kick in.

e.g.

val x: String|Null = ???
if (x == null) return
def foo() = bar()
def bar() = x.length // won't compile because we use the creation context (due to the forward reference from `foo`)